server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/SearchAction.java

```diff
@@ -0,0 +1,159 @@
+import org.sonar.api.user.User;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.component.ComponentDto;
+import org.sonar.core.measure.custom.db.CustomMeasureDto;
+import org.sonar.core.metric.db.MetricDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.es.SearchOptions;
+import org.sonar.server.user.index.UserIndex;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class SearchAction implements CustomMeasuresWsAction {
+
+  public static final String ACTION = "search";
+  public static final String PARAM_PROJECT_ID = "projectId";
+  public static final String PARAM_PROJECT_KEY = "projectKey";
+
+  private final DbClient dbClient;
+  private final UserIndex userIndex;
+  private final CustomMeasureJsonWriter customMeasureJsonWriter;
+
+  public SearchAction(DbClient dbClient, UserIndex userIndex, CustomMeasureJsonWriter customMeasureJsonWriter) {
+    this.dbClient = dbClient;
+    this.userIndex = userIndex;
+    this.customMeasureJsonWriter = customMeasureJsonWriter;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setDescription("List custom measures. The project id or project key must be provided.")
+      .setSince("5.2")
+      .addFieldsParam(CustomMeasureJsonWriter.OPTIONAL_FIELDS)
+      .addPagingParams(100)
+      .setResponseExample(getClass().getResource("example-search.json"))
+      .setHandler(this);
+
+    action.createParam(PARAM_PROJECT_ID)
+      .setDescription("Project id")
+      .setExampleValue("ce4c03d6-430f-40a9-b777-ad877c00aa4d");
+
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("Project key")
+      .setExampleValue("org.apache.hbas:hbase");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    String projectUuid = request.param(PARAM_PROJECT_ID);
+    String projectKey = request.param(PARAM_PROJECT_KEY);
+    List<String> fieldsToReturn = request.paramAsStrings(WebService.Param.FIELDS);
+    SearchOptions searchOptions = new SearchOptions()
+      .setPage(request.mandatoryParamAsInt(WebService.Param.PAGE),
+        request.mandatoryParamAsInt(WebService.Param.PAGE_SIZE));
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto project = searchProject(dbSession, projectUuid, projectKey);
+      List<CustomMeasureDto> customMeasures = dbClient.customMeasureDao().selectByComponentUuid(dbSession, project.uuid(), searchOptions);
+      int nbCustomMeasures = dbClient.customMeasureDao().countByComponentUuid(dbSession, project.uuid());
+      Map<String, User> usersByLogin = usersByLogin(customMeasures);
+      Map<Integer, MetricDto> metricsById = metricsById(dbSession, customMeasures);
+      JsonWriter json = response.newJsonWriter();
```

Comment:
split content of try catch into smaller method to gain in clarity and make intent more explicit